Diffuser change upgraded to 0.26.3 along with MLPERF SD XL support

ANSHUMAN87 opened the pull request:
1. MLPERF SD XL inference related change unstreamed
    2. All Diffuser changes upgraded to 0.26.3
JH...

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_1135). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sywangyi commented:
@yuanwu2017 @cfgfung please be ware of it.

dsocek commented on optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:
```diff
@@ -34,15 +38,629 @@
+                                            callback_steps, ip_adapter_image,
+                                            image_embeds, timestep_cond, callback_on_step_end,
+                                            callback_on_step_end_tensor_inputs)
+                for i, t in enumerate(timesteps[-2:], 18):
```
Why hard code 18, do we assume 20 timesteps/inference steps?

ANSHUMAN87 replied:
Yes there will be only 20 steps/inference in case of MLPERF SD XL .

JH...!

sywangyi commented on optimum/habana/diffusers/models/attention_processor.py:
```diff
@@ -0,0 +1,188 @@
+          super().__init__()
+
+      def forward(self, x, dim = None, invAttnHead= None):
+          return torch.ops.hpu.softmax_fp8(x, dim,  None, None, invAttnHead)
```
why is softmax_fp8 used here? support stable diffuser for fp8?

ANSHUMAN87 replied:
Current support is for FP8 only.

sywangyi commented on optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:
```diff
@@ -34,15 +38,629 @@
+    set_attn_processor_hpu(self, processor)
+
+
+class StableDiffusionXLPipeline_HPU(StableDiffusionXLPipeline):
```
StableDiffusionXLPipeline_HPU is not used in any example.

ANSHUMAN87 replied:
StableDiffusionXLPipeline_HPU implementation is used by MLPERF SD XL workload published in model garden. So the usage in optimum-habana wont be visible. 

JH...!

regisss replied:
Let's move this new class and all the related changes in this file to a new file called `pipeline_stable_diffusion_xl_mlperf.py` please. It will be clearer and easier to maintain.

sywangyi commented on optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py:
```diff
@@ -224,19 +230,33 @@ def step(
         sample = sample.to(torch.float32)
 
-        sigma, sigma_next = self.get_params(timestep)
+        if self.hpu_opt:
```
what's is this hpu_opt for? see **self.hpu_opt = False** in __init__, not found "self.hpu_opt = True" in any place.

ANSHUMAN87 replied:
"self.hpu_opt = True" will be done in MLPERF SD XL repo where we instantiate the class. In optimum-habana, wont be visible. By default is set to Flase, so that current flow is intact in optimum-habana.

JH...!

libinta replied:
@ANSHUMAN87 do you mean optimum-habana doesn't need hpu_opt =true and no need for optimum-habana ? or you don't have method to set it to true for optimum-habana?

ANSHUMAN87 replied:
As of now, optimum-habana no need to set to True, in case in future needed, user can instantiate the class and set to True, if user wants to use those optimizations. Those optimizations are applicable only for MLPERF SD-XL repo currently.

libinta commented on optimum/habana/diffusers/models/attention_processor.py:
```diff
@@ -0,0 +1,188 @@
+        self.bmm1 = Matmul()
+        self.bmm2 = Matmul()
+        self.softmax = Softmax()
+
```
@ANSHUMAN87 the scoped optimization for matmu;/softmax is for fp8? and how to invoke it for bf16 model run?

kumarans-ai replied:
@libinta , we have both options when you pass scale as None automatically kernel will operate in bf16. Please refer softmax API for further

dsocek commented on optimum/habana/diffusers/models/unet_2d_condition.py:
```diff
@@ -1,349 +1,352 @@
-
-        if self.config.class_embeddings_concat:
-            emb = torch.cat([emb, class_emb], dim=-1)
+        self,
```
Can you, please, fix indentation?   (remove 4 spaces from all lines in this function)

dsocek commented on optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:
```diff
@@ -570,7 +1188,7 @@ def __call__(
 
             # 4. Prepare timesteps
-            timesteps, num_inference_steps = retrieve_timesteps(self.scheduler, num_inference_steps, device, timesteps)
+            timesteps, num_inference_steps = retrieve_timesteps_v2(self.scheduler, num_inference_steps, device, timesteps)
```
Can you put back the original name, and rename the `retrieve_timesteps` specific to ML perf _hpu class instead?

dsocek reviewed (approved):
LGTM as long as OK with CODEOWNERS to add an extra MLPerf-specific class to OH

libinta commented:
@ANSHUMAN87  @dsocek what are the change needed to make sure we can utilize the mlperf optimization?

yeonsily commented:
@ANSHUMAN87 @dsocek can you please check Libin's comment and also fix code style?

dsocek commented:
> @ANSHUMAN87 @dsocek can you please check Libin's comment and also fix code style?

@libinta, @yeonsily I think that @ANSHUMAN87 is better to answer this as he created the code/PR.  We can consider if some of his optimization strategies can be applied to a normal OH Stable Diffusion pipeline (but this means no hard coded params in pipeline as users may select/run different use-cases)

ANSHUMAN87 commented:
> @ANSHUMAN87 @dsocek what are the change needed to make sure we can utilize the mlperf optimization?

We have MLPERF specific optimizations only in 2 places, other places are combined for both:
1. Euler scheduler (optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py): User need to set **hpu_opt** field to True explicitly after instantiation of **GaudiEulerDiscreteScheduler** class.
2. Stable Diffusion XL pipeline (optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py): User need to use **StableDiffusionXLPipeline_HPU** class instead of **GaudiStableDiffusionXLPipelineOutput**

Please let me know, if need any more info.

JH...!

ANSHUMAN87 commented:
> @ANSHUMAN87 @dsocek can you please check Libin's comment and also fix code style?

@yeonsily I had run ruff in my local, all errors are handled already. But the import error reported in this PR, I am unable to reproduce in my local. Any suggestions ?

JH...!

dsocek commented:
> > @ANSHUMAN87 @dsocek what are the change needed to make sure we can utilize the mlperf optimization?
> 
> We have MLPERF specific optimizations only in 2 places, other places are combined for both:
> 
> 1. Euler scheduler (optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py): User need to set **hpu_opt** field to True explicitly after instantiation of **GaudiEulerDiscreteScheduler** class.
> 2. Stable Diffusion XL pipeline (optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py): User need to use **StableDiffusionXLPipeline_HPU** class instead of **GaudiStableDiffusionXLPipelineOutput**
> 
> Please let me know, if need any more info.
> 
> JH...!

@ANSHUMAN87 Thanks, but more info is needed here.  Can you get into the details of the strategies in **StableDiffusionXLPipeline_HPU** class, and what is different when compared to general class **GaudiStableDiffusionXLPipeline**.

ANSHUMAN87 commented:
> > > @ANSHUMAN87 @dsocek what are the change needed to make sure we can utilize the mlperf optimization?
> > 
> > 
> > We have MLPERF specific optimizations only in 2 places, other places are combined for both:
> > 
> > 1. Euler scheduler (optimum/habana/diffusers/schedulers/scheduling_euler_discrete.py): User need to set **hpu_opt** field to True explicitly after instantiation of **GaudiEulerDiscreteScheduler** class.
> > 2. Stable Diffusion XL pipeline (optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py): User need to use **StableDiffusionXLPipeline_HPU** class instead of **GaudiStableDiffusionXLPipelineOutput**
> > 
> > Please let me know, if need any more info.
> > JH...!
> 
> @ANSHUMAN87 Thanks, but more info is needed here. Can you get into the details of the strategies in **StableDiffusionXLPipeline_HPU** class, and what is different when compared to general class **GaudiStableDiffusionXLPipeline**.

Below are the high level changes done in STableDiffusionXLPipeline_HPU:
   - Added mark_step after unet Model in the pipeline
   - Added attention_processor implementation
   - Added unet forward implementation
   - Added quantization support

JH...!

regisss commented on optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:
```diff
@@ -34,16 +38,714 @@
     CLIPVisionModelWithProjection,
 )
 
+from optimum.habana.diffusers import GaudiDiffusionPipeline
+from optimum.habana.diffusers.models.attention_processor import (
+    AttentionProcessor,
+    AttnProcessor2_0,
+    ScaledDotProductAttention,
+)
+from optimum.habana.diffusers.models.unet_2d_condition import gaudi_unet_2d_condition_model_forward
+from optimum.habana.diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion import (
+    retrieve_timesteps,
+)
+from optimum.habana.transformers.gaudi_configuration import GaudiConfig
+from optimum.habana.utils import HabanaProfile, speed_metrics, warmup_inference_steps_time_adjustment
```
```suggestion
from ...diffusers import GaudiDiffusionPipeline
from ...diffusers.models.attention_processor import (
    AttentionProcessor,
    AttnProcessor2_0,
    ScaledDotProductAttention,
)
from ...models.unet_2d_condition import gaudi_unet_2d_condition_model_forward
from ..pipelines.stable_diffusion.pipeline_stable_diffusion import (
    retrieve_timesteps,
)
from ....transformers.gaudi_configuration import GaudiConfig
from ....utils import HabanaProfile, speed_metrics, warmup_inference_steps_time_adjustment
```
Please use relative imports in the source code

regisss reviewed (commented):
I left a few comments, the main one being about moving all the code added to `pipeline_stable_diffusion_xl.py` to a new file so that the codebase is clearer and easier to maintain.

ANSHUMAN87 commented:
@regisss your comments are addressed now. Thanks!

JH...!

dsocek commented on optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_mlperf.py:
```diff
@@ -0,0 +1,739 @@
+        self._num_timesteps = len(timesteps)
+        with self.progress_bar(total=num_inference_steps) as progress_bar:
+            timesteps = [t.item() for t in timesteps]
+            if self.quantized:
```
Where is `quantized` defined for the class?

dsocek commented on optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_mlperf.py:
```diff
@@ -0,0 +1,739 @@
+        force_zeros_for_empty_prompt: bool = True,
+        add_watermarker: Optional[bool] = None,
+    ):
+        super().__init__(
```
Why are we not inheriting gaudi-specific class members/configs from `GaudiDiffusionPipeline`?

ANSHUMAN87 replied:
@dsocek I did not find any specific need to use GaudiDiffusionPipeline. But I am open for suggestions.

imangohari1 commented:
@ANSHUMAN87 
Hi, 
Could you rebase and re-test this with the main? The main diffuser is now that 0.29.2 and some changes are needed for this PR. Thanks.

regisss commented:
Closing as these changes were added to https://github.com/huggingface/optimum-habana/pull/1204.
